File: rd-net/Test.RdFramework/Reflection/SerializersTest.cs
using System.Collections.Generic;
using JetBrains.Diagnostics;
using JetBrains.Rd.Impl;
using JetBrains.Serialization;
using NUnit.Framework;

namespace Test.RdFramework.Reflection
{
  [TestFixture]
  public class SerializersTest
  {
    [Test]
    public unsafe void TestReadRName()
    {
      var testNames = new List<RName> {
        RName.Empty,
        RName.Empty.Sub("", ""),
        RName.Empty.Sub("abc", ""),
        RName.Empty.Sub("some very long string with numbers 1234567890 and strange \u0d78\u0bf5 symbols", ""),
        RName.Empty.Sub("abc", "").Sub("asdf123", "::"),
        RName.Empty.Sub("arbitrary", "").Sub("separators with", " spaces and $&*@ symbols "),
        RName.Empty.Sub("a", "").Sub("b", ".").Sub("c", "::").Sub("d", "$").Sub("e", "_").Sub("$", ".").Sub("[]", "::"),
        RName.Empty.Sub("", "").Sub("", "").Sub("", "").Sub("", "").Sub("", "")
      };

      using var cookie = UnsafeWriter.NewThreadLocalWriter();
      var writer = cookie.Writer;
      foreach (var name in testNames)
      {
        var start = writer.Ptr;
        ExtCreatedUtils.WriteRName(writer, name);
        var reader = UnsafeReader.CreateReader(start, 1000);
        var value = ExtCreatedUtils.ReadRName(reader);

        Assert.True(RNameEquals(name, value), $"expected \"{name}\" but got \"{value}\"");
      }
    }
    
    private bool RNameEquals(RName a, RName b)
    {
      if (a == RName.Empty || b == RName.Empty)
        return a == b;
      if (a.LocalName.ToString() != b.LocalName.ToString() || a.Separator != b.Separator)
        return false;
      if (!(a.Parent is RName aParent) || !(b.Parent is RName bParent))
        return a == b;
      return RNameEquals(aParent, bParent);
    }
  }
}

File: rd-kt/rd-framework/src/test/kotlin/com/jetbrains/rd/framework/test/cases/serialization/SerializersTest.kt
package com.jetbrains.rd.framework.test.cases.serialization

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.util.string.RName
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class SerializersTest {

    @Test
    fun read_date_should_floor_negative_millis() {
        val buffer = UnsafeBuffer(ByteArray(100))
        buffer.writeLong(621355950301790001) // 1969-12-31 23:30:30.1790001 in ticks
        buffer.rewind()

        val dateTime = buffer.readDateTime()

        assertEquals (dateTime.time, -1769821) // 1969-12-31 23:30:30.179, NOT 1969-12-31 23:30:30.180
    }

    @Test
    fun testReadArray() {
        val buffer = UnsafeBuffer(ByteArray(10))
        buffer.writeArray(Array(1) {"abc"}) {}
        buffer.rewind()

        assertTrue (arrayOf("abc") contentEquals buffer.readArray { "abc" })
    }

    @Test
    fun testReadRName() {
        val testNames = listOf(
            RName.Empty,
            RName.Empty.sub("", ""),
            RName.Empty.sub("abc", ""),
            RName.Empty.sub("some very long string with numbers 1234567890 and strange \u0d78\u0bf5 symbols", ""),
            RName.Empty.sub("abc", "").sub("asdf123", "::"),
            RName.Empty.sub("arbitrary", "").sub("separators with", " spaces and $&*@ symbols "),
            RName.Empty.sub("a", "").sub("b", ".").sub("c", "::").sub("d", "$").sub("e", "_").sub("$", ".").sub("[]", "::"),
            RName.Empty.sub("", "").sub("", "").sub("", "").sub("", "").sub("", "")
        )

        val buffer = UnsafeBuffer(ByteArray(1000))
        for (name in testNames) {
            buffer.rewind()
            RNameMarshaller.write(buffer, name)
            buffer.rewind()
            val value = RNameMarshaller.read(buffer)

            assertTrue(rNameEquals(name, value)) { 
                "expected \"$name\" but got \"$value\""
            }
        }
    }
    
    
    private fun rNameEquals(a: RName, b: RName): Boolean {
        if (a == RName.Empty || b == RName.Empty)
            return a == b
        if (a.localName != b.localName || a.separator != b.separator)
            return false
        return rNameEquals(a.parent ?: RName.Empty, b.parent ?: RName.Empty)
    }
}